sdk/python/kfp_tekton/compiler/compiler.py

```diff
@@ -463,6 +464,138 @@ def _create_dag_templates(self, pipeline, op_transformers=None, params=None, op_
+    def _get_inputs_outputs_recursive_opsgroup(group):
+      # TODO: refactor the following codes with the above
+      if group.recursive_ref:
+        params = [(param, False) for param in group.inputs]
+        params.extend([(param, True) for param in list(condition_params[group.name])])
+        for param, is_condition_param in params:
+          if param.value:
+            continue
+          full_name = param.full_name
+          if param.op_name:
+            upstream_op = pipeline.ops[param.op_name]
+            upstream_groups, downstream_groups = \
+              self._get_uncommon_ancestors(op_groups, opsgroup_groups, upstream_op, group)
+            for i, g in enumerate(downstream_groups):
+              if i == 0:
+                inputs[g].add((full_name, upstream_groups[0]))
+              # There is no need to pass the condition param as argument to the downstream ops.
+              # TODO: this might also apply to ops. add a TODO here and think about it.
+              elif i == len(downstream_groups) - 1 and is_condition_param:
+                continue
+              else:
+                # For Tekton, do note append duplicated input parameters
+                duplicated_downstream_group = False
+                for group_name in inputs[g]:
+                  if len(group_name) > 1 and group_name[0] == full_name:
+                    duplicated_downstream_group = True
+                    break
+                if not duplicated_downstream_group:
+                  inputs[g].add((full_name, None))
+            for i, g in enumerate(upstream_groups):
+              if i == len(upstream_groups) - 1:
+                outputs[g].add((full_name, None))
+              else:
+                outputs[g].add((full_name, upstream_groups[i + 1]))
+          elif not is_condition_param:
+            for g in op_groups[group.name]:
+              inputs[g].add((full_name, None))
+      for subgroup in group.groups:
+        _get_inputs_outputs_recursive_opsgroup(subgroup)
+
+    _get_inputs_outputs_recursive_opsgroup(root_group)
+
+    # Generate the input for SubGraph along with parallelfor
+    for sub_graph in opsgroup_groups:
+      if sub_graph in op_name_to_for_loop_op:
+        # The opsgroup list is sorted with the farthest group as the first and
+        # the opsgroup itself as the last. To get the latest opsgroup which is
+        # not the opsgroup itself -2 is used.
+        parent = opsgroup_groups[sub_graph][-2]
+        if parent and parent.startswith('subgraph'):
+          # propagate only op's pipeline param from subgraph to parallelfor
+          loop_op = op_name_to_for_loop_op[sub_graph]
+          pipeline_param = loop_op.loop_args.items_or_pipeline_param
+          if loop_op.items_is_pipeline_param and pipeline_param.op_name:
+            param_name = '%s-%s' % (
+              sanitize_k8s_name(pipeline_param.op_name), pipeline_param.name)
```

Comment:
So here the only difference if to sanitize the parameter name? In fact I think we do not really support subgraph now but it's of no harm.